stage0/src/sev.rs

```diff
@@ -104,18 +108,221 @@ pub fn deinit_ghcb(snp: bool, encrypted: u64) {
     tlb::flush_all();
 }
 
+// Mapper between the page sizes in the instructions module doesn't use the PageSize from the x86_64
+// crate. The main reason for that is that you can't call PVALIDATE on 1 GiB pages.
+pub trait ValidatablePageSize {
+    const SEV_PAGE_SIZE: SevPageSize;
+}
+
+impl ValidatablePageSize for Size4KiB {
+    const SEV_PAGE_SIZE: SevPageSize = SevPageSize::Page4KiB;
+}
+
+impl ValidatablePageSize for Size2MiB {
+    const SEV_PAGE_SIZE: SevPageSize = SevPageSize::Page2MiB;
+}
+
+trait Validate<S: PageSize> {
+    fn pvalidate(&self) -> Result<(), InstructionError>;
+}
+
+impl<S: PageSize + ValidatablePageSize> Validate<S> for Page<S> {
+    fn pvalidate(&self) -> Result<(), InstructionError> {
+        pvalidate(
+            self.start_address().as_u64() as usize,
+            S::SEV_PAGE_SIZE,
+            Validation::Validated,
+        )
+    }
+}
+
+/// Container for a page of virtual memory, and the page table that controls the mappings for that
+/// page.
+struct MappedPage<S: PageSize> {
+    pub page: Page<S>,
+    pub pt: PageTable,
+}
+
+impl<S: PageSize> MappedPage<S> {
+    pub fn new(vaddr: VirtAddr) -> Result<Self, AddressNotAligned> {
+        let mapped_page = Self {
+            page: Page::from_start_address(vaddr)?,
+            pt: PageTable::new(),
+        };
+        Ok(mapped_page)
+    }
+}
+
+fn pvalidate_range<S: PageSize + ValidatablePageSize, T: PageSize, F>(
+    range: &PhysFrameRange<S>,
+    memory: &mut MappedPage<T>,
+    encrypted: u64,
```

Comment:
What is this? Should this be renamed to something like length or address?

Reply:
This is the location of the encrypted bit in the page table.

For AMD Milan, if a page is to be encrypted, we need to set the 51-th bit in the page table. This value is impementation-defined, so we carry around this u64 to denote what needs to be set (or cleared) for encryption.

Reply:
I see. Thanks for clarifying. I assume it is sufficiently documented somewhere :)